### ISSUE TITLE
Question: Multiple singleuser servers

### ISSUE THREAD

**Issue by danielfrg** [jupyterhub/jupyterhub#766]
Is there plans (or interest) to add support for multiple single-user servers per user? I kinda see it in the [roadmap project sharing](https://github.com/jupyter/roadmap#next--1-08) but not sure how is that planned on being implemented, from what I can guess is basically multiple users pushing to the same git repo that is managed by one service in JupyterHub.

Is there interest on refactoring the code to support this feature? Looking at the models in `orm.py` right now the one single-user server is a little hardcoded in the `User` class but I guess that this is something that can change.

Another solution is probably to have "admin" servers that multiple users can access based on some permissions.

Basically I am just asking for thoughts on these ideas and if there is interest on those features to be added upstream in case they get implemented.

**Comment by willingc** [comment]
Thanks @danielfrg for your thoughts and questions about the roadmap. @minrk would be the best person to provide insights into planning for release 0.8+. :sunny:

**Comment by micha-f** [comment]
We are also interested in this. In our case we'd like to use the docker spawner to spawn multiple containers per user - with different settings per notebook. We'd be happy to contribute if there is an agreement on whether and how to implement support for multiple servers per user.

**Comment by minrk** [comment]
This is definitely planned. Changing server from a singleton to a list in the ORM is not very complicated. The challenge is more in the UI and REST API. We would need:

1. UI for listing and managing servers, rather than a single stop/start buttons.
2. Same thing for the admin page (a little more complex, since it's much more information)
3. last_activity needs to be added as a server-level data point
4. make some cookie/URL-related decisions, e.g.
    - single cookie per user
    - user servers at `/user/:name/:server-id`

And, with all of this, we want deployments to be able to choose to preserve the single-server-per-user notion, which dramatically simplifies the application for users (the Hub is just a login page to them, in this case). This mostly affects views and not the API, but it could reach into the API a little bit in a few places.

**Comment by parente** [comment]
We're also interested in this for giving users the ability to run classic notebook and new jupyter lab separately. (Running both out of the same server instance has some sideeffects like clicking the Jupyter logo always returning the user to /lab even when they were in a classic notebook). My stopgap is to bring up a second jupyterhub instance, but a single hub instance that can manage multiple servers is clearly better in the long run.

I can help out some here.

**Comment by barrachri** [comment]
Hello !

can we set a kind of roadmap ?

I plan to submit a PR decoupling the User and Server models inside orm.py

`_server_id = Column(Integer, ForeignKey('servers.id', ondelete="SET NULL"))`
`server = relationship(Server, primaryjoin=_server_id == Server.id)`

To mantain a compatibility with one-server-per-user _server_id in the new version can be a property where you get the first (and only) server.
server can remain the same limiting the result to just one entry and we can add servers as a backref to all the user's servers.

Same for last_activity.

hints & thoughts ?

**Comment by willingc** [comment]
@barrachri You can find the current JupyterHub roadmap here: https://github.com/jupyter/roadmap/blob/master/jupyterhub.md

@minrk Thoughts on the proposed change?

**Comment by minrk** [comment]
@barrachri sounds great! I don't think we need to keep `_server_id`, I think only `server` is needed as a property mapping to the first server (if any, None otherwise), and we can drop `_server_id`. The user last_activity should stay as it is. We will need to *add* a last_activity field to servers, but we will probably still want a value on the user itself (this is affected by API requests, etc.

**Comment by barrachri** [comment]
what about mapping last_activity (inside the user) to last_activity of _server_id ?

Also having multi-servers makes sense, but it makes even more sense to have multi-configuration for the servers.
Regarding docker we have this value inside jupyterconfig_hub.py: c.DockerSpawner.container_image = os.environ['DOCKER_NOTEBOOK_IMAGE']

I think would be great to add a parameter to the API to choose which image to use and in case of None use container_image as default.

And yes, I think as a consequence we should add this inside the Spawner.

**Comment by minrk** [comment]
> I think would be great to add a parameter to the API to choose which image to use and in case of None use container_image as default.

This is already supported via the `options_form` configuration, and used by spawners such as ProfileSpawner and WrapperSpawner. The same can be done on DockerSpawner, if a list of container images is given. Allowing *any* image is likely not something that DockerSpawner should do by default.

**Comment by barrachri** [comment]
Thanks @minrk.

I am checking the JupyterHub API, for example POST /users/{name}/server to create a new server

I see it doesn't accept any parameter and that it doesn't return nothing.

That said I would like to prepare a PR with:

1. accepting a json body with the same use of options_form
2. return some info about the server
3. add the possibility to create multiple servers
4. if the token is registered as "admin" have the possibility to receive the cookie to access the user's server
5. other ?

**Comment by minrk** [comment]
@barrachri great! I think the doc might be a bit out of date, because the REST API *does* support user_options as the JSON body.

1. This is already supported. Make a POST with json body, and it will be interpreted as `user_options`. We may need to change this if we need to expose new options, pushing user_options down a level in the data structure.
2. A JSON response is a great idea. It's empty now, because there's only one place a user's server can go. Multiple servers changes this.
3. Indeed. We'll need to be careful, because most of the time requesting a server *shouldn't* create a  new server if one is running / pending launch. Perhaps we need a new endpoint such as `POST /user/:name/server/:name` for creating new named servers.
4. I'd rather keep separate admin access requests from launch. I think that's doing too much at a time.

**Comment by barrachri** [comment]
Point n. 4 is more if you want to use the JupyterHub API instead of the UI to give access to the user's servers.

If you want for example to create a new server and give access to that server to the user how can you do this without using the login form ?

**Comment by minrk** [comment]
> If you want for example to create a new server and give access to that server to the user how can you do this without using the login form?

- `POST /user/:name/server` starts the server
- `POST /user/:name/admin-access` sets the cookie for admin-access to the user's server

We'll need to reconsider some of these endpoints when there are multiple servers, as `admin-access` will need to be per-server, not per-user, which used to be the same thing.

**Comment by barrachri** [comment]
I tried 'POST /user/:name/admin-access'

But it doesn't work with the services and it seems more a browser endpoint instead of a REST API.
Also I don't receive any response.

**Comment by minrk** [comment]
Admin access means setting a cookie that the single-user server will see, so in that sense it is mainly useful in browsers. But it is part of the REST API, and used as such:

```
# start minrk's server
$ curl -X POST -H "Authorization: token $TOKEN" http://127.0.0.1:8765/hub/api/users/minrk/server
# request admin-access
$ curl -X POST -v -H "Authorization: token $TOKEN" http://127.0.0.1:8765/hub/api/users/minrk/admin-access
...
< HTTP/1.1 200 OK
< set-cookie: jupyter-hub-token-minrk="2|1:0|10:1484318029|23:jupyter-hub-token-minrk|xxxxxxxxxxxxxxx"; expires=Sun, 12 Feb 2017 14:33:49 GMT; Path=/user/minrk
...
```

The response is:

1. 200 OK (access granted)
2. setting the cookie needed to access the server

**Comment by barrachri** [comment]
@minrk thanks for the patience, I was using Postman and there is not output there :/

**Comment by barrachri** [comment]
Some questions:

Is user_options generally sent down to the spawner as a dict ?
What do we want to use for server name in  `/user/:name/server/:name` ? the id ?

Right now the cookies are built is this way: name(user.server.cookie_name): value(user.cookie_id)
At the end you have something like jupyter-hub-token-christianbarra: 39489sfjkdhfdhjhj4h3j4h3j4j

do we want to keep the same ?

**Comment by minrk** [comment]
Follow-up from gitter:

> Is user_options generally sent down to the spawner as a dict ?

user_options is set as the `.user_options` attribute on the Spawner. It is up to the spawner to interpret these options.

> What do we want to use for server name in /user/:name/server/:name ? the id ?

It should probably be specified by the request that creates it (i.e. names are required). If names are not required, we could use counters by default - 1, 2, 3, etc. or simple random hex IDs.

> Right now the cookies are built is this way

We will probably want to keep setting cookies for `/user/:name` and use the same cookie for all of the user's servers under that prefix. That makes the base url prefix, cookie name and ID a property of the user, and *not* a property of the user's server.

**Comment by barrachri** [comment]
for server :name we can use
`uuid.uuid4()` and then `base64.urlsafe_b64encode(uuid)`

For the cookie you said it is better to use one cookie per user.

**Comment by barrachri** [comment]
What is left to complete the named servers part ?

**Comment by minrk** [comment]
We'll need UI for the multi-server case. Users are going to need to be able to list and manage their own servers.

We're also going to have to think carefully about what happens when a user first logs in, because right now that spawns their default server, which it maybe should not in the named server case.

Once we've figured out all of the database additions, we are also going to need an alembic upgrade file.

**Comment by barrachri** [comment]
> We're also going to have to think carefully about what happens when a user first logs in, because > right now that spawns their default server, which it maybe should not in the named server case.

Thoughts about this ?

in case you use named_severs and you go to /user/{YourUser}
should you get an error or not ?

Do we need to think about the cookies ?

**Comment by minrk** [comment]
> Do we need to think about the cookies ?

No, for two reasons:

1. The cookies are set on `/user/name`, which means that all the single-user servers below there will share the same cookie.
2. I hope to switch over to OAuth internally by the time we release 0.8, in which case the cookie paths won't be relevant anymore

**Comment by minrk** [comment]
> Thoughts about this ?
> in case you use named_severs and you go to /user/{YourUser}
> should you get an error or not ?

My current thought is that the default page and `/user/:name` send you to a (new) page where you can see your list of named servers and request new ones.

**Comment by willingc** [comment]
@minrk FYI Ben Zaitlen has been doing quite a bit of work on OAuth for CA.

**Comment by barrachri** [comment]
> My current thought is that the default page and /user/:name send you to a (new) page where you can > see your list of named servers and request new ones.

It makes sense, I would also give the possibility to redirect /user/:name to something else, maybe you can define this inside the configuration.

> We're also going to have to think carefully about what happens when a user first logs in, because > right now that spawns their default server, which it maybe should not in the named server case.

I guess the best thing is just to redirect the user to the user page without spawning a server.

**Comment by barrachri** [comment]
> I hope to switch over to OAuth internally by the time we release 0.8, in which case the cookie paths won't be relevant anymore

**Comment by dhirschfeld** [comment]
One thing which would be useful is to be able to give different groups of users different options for the servers (containers)

**Comment by minrk** [comment]
The internal implementation and REST API of named servers has landed now in #1243. We still need UI pages for this, but I don't think we'll get those in 0.8, which we are trying to finalize.

**Comment by jpays** [comment]
Hi, 
This feature would be excessively useful to us, any chance it could be prioritized again ?
Our only other option is launching several jupyterhub instances, which is not ideal.
Thanks

**Comment by Carreau** [comment]
There is already the `--/c.JupyterHub.allow_named_servers=True` whcih allow one user to run multipel servers. I'm unsure if it's possible to start the servers with different profiles.

**Comment by rcthomas** [comment]
We just managed to do this thanks especially to #2435.  We created a custom home/console, and then we use a spawner based on wrapspawner to manage the profiles.  We'll work on documenting it for others who may find it useful.

**Comment by realAndyLuo** [comment]
@rcthomas It's awesome to know you managed to do this. 
Any chance you have finished documenting it?

**Comment by zeehio** [comment]
With dockerspawner (or SystemUserSpawner) it is very easy to set up. @oulydna see https://github.com/jupyterhub/dockerspawner/issues/257#issuecomment-492275806

```
c.JupyterHub.allow_named_servers=True
c.JupyterHub.spawner_class = 'dockerspawner.SystemUserSpawner'
c.SystemUserSpawner.name_template = '{prefix}-{username}-{servername}'
c.SystemUserSpawner.image_whitelist = ['jupyter/scipy-notebook', 'jupyter/tensorflow-notebook']
```

**Comment by manics** [comment]
This was implemented a while ago: https://jupyterhub.readthedocs.io/en/1.1.0/reference/config-user-env.html#named-servers
